File: Hash map/409_LongestPalindrome.cs
// 409. Longest Palindrome
// Tags: hash map
//
public class Solution 
{
    public int LongestPalindrome(string s) 
    {
        // Edge cases + optimiztion.
        if(string.IsNullOrEmpty(s))
            return 0;
        else if(s.Length == 1)
            return 1;
        
        int result = 0;
        Dictionary<char, int> dict = new();
        
        // Store all values in hash map.
        foreach(char c in s)
        {
            if(!dict.ContainsKey(c))
                dict.Add(c, 1);
            else
                dict[c] += 1;
        }
        
        // Count all the even lengthed and odd lengthed 
        // (greater than or equal to 3) entries.
        bool oddEncountered = false;
        foreach(KeyValuePair<char, int> i in dict)
        {
            // Even numbers.
            if(i.Value % 2 == 0)
            {
                result += i.Value;
            }
            // Odd numbers.
            else
            {
                oddEncountered = true;
                result += i.Value - 1;
            }
        }
        
        // If we had any odd number entries, add 1 to the middle
        // of our palindrome.
        if(oddEncountered)
            result++;
        
        return result;
    }
}

File: Hash map/409_LongestPalindrome.rs
// 409. Longest Palindrome
// Tags: hash map, closures
//

// Most credit goes to:
// https://leetcode.com/problems/longest-palindrome/solutions/1949080/rust-solution/
impl Solution {
    pub fn longest_palindrome(s: String) -> i32 {
        let mut dict = std::collections::HashMap::new();

        // Setup our hash map.
        s.chars().for_each(|c| {
            *dict.entry(c).or_insert(0) += 1;
        });

        // Count all the even lengthed and odd lengthed 
        // (greater than or equal to 3) entries.
        let mut count = dict.values().fold(0, |acc, &v| {
            acc + match v % 2 == 0 {
                true => v,
                false => v - 1, // We remove one number from our odd count so as to keep the palindrome valid for the timebeing.
            }
        });

        // If we have remove, add another character (one of our odd characters that were removed)
        // to the final answer.
        if count < s.len() {
            count += 1;
        }

        count as i32
    }
}